lib/ansible/modules/crypto/certificate_complete_chain.py

```diff
@@ -0,0 +1,355 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# (c) 2018, Felix Fontein <felix@fontein.de>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: certificate_complete_chain
+author: "Felix Fontein (@felixfontein)"
+version_added: "2.7"
+short_description: Complete certificate chain given a set of untrusted and root certificates
+description:
+    - "This module completes a given chain of certificates in PEM format by finding
+       intermediate certificates from a given set of certificates, until it finds a root
+       certificate in another given set of certificates."
+    - "This can for example be used to find the root certificate for a certificate chain
+       returned by M(acme_certificate)."
+    - "Note that this module does I(not) check for validity of the chains. It only
+       checks that issuer and subject match, and that the signature is correct. It
+       ignores validity dates and key usage completely. If you need to verify that a
+       generated chain is valid, please use C(openssl verify ...)."
+requirements:
+    - "cryptography >= 1.5"
+options:
+    input_chain:
+        required: no
+        description:
+            - A concatenated set of certificates in PEM format forming a chain.
+            - The module will try to complete this chain.
+    input_chain_src:
```

Comment:
not a blocker (IMHO) but having 2 params for input chain: `input_chain_src` and `input_chain` is not necessary (I know acme_certificate has also 2 but this has historical reasons). I would use only `input_chain`, because reading from a file could be done with a lookup plugin: `input_chain: "{{ lookup('file', '/path/filename' )}". Also remember, lookup plugins can also "read" from all kinds of things, like DBs, Vaults, kv stores.

Reply:
Ah, that's a good point. I'll change the module accordingly, since here the use-case "input data is on remote system" isn't very important (and the input files are small).

One question though: what happens if the contents of a file on the remote system are needed? With `input_chain_src`, this is easily doable, but without it? Do I have to do a `fetch: ...` first in that case, before I can use this module?

Reply:
it would require to have the content into a var e.g like with the slurp module.

Reply:
how would this work with `input_chain_src`?

Reply:
Ah sorry, I meant `slurp: ...`, not `fetch: ...`. And yes, just having `..._src` makes this complicated.

But: for `acme_certificate` (and friends), I think it is different, and both parameters should be provided. The reason is that depending on how you use the module, and how your security model is set up, you do *not* want the private account key to leave the remote host where the module is executed. If you have to `slurp` it first, this condition is violated. So there, having both `account_key_content` and `account_key_src` is a Very Good Thing(tm).

But that doesn't apply to this module, since certificates are usually public enough. Also, the input size isn't huge (another potential reason for having both parameters).

Reply:
@remove I've removed `input_chain_src` in 681a468ba619f0e9630b8caa29dd849d1691d715.